src/query/storages/fuse/src/operations/replace_into/mutator/mutator_replace_into.rs

```diff
@@ -12,58 +12,160 @@
+        metrics_inc_replace_row_number_after_table_level_pruning(row_number_after_pruning as u64);
+
+        if row_number_after_pruning == 0 {
+            info!("(replace-into) all rows are append-only");
+            return Ok(MergeIntoOperation::None);
+        }
+
+        let merge_into_operation = if let Some(partitioner) = &self.partitioner {
+            // if table has cluster keys; we partition the input data block by left most column of cluster keys
+            let partitions = partitioner.partition(data_block)?;
+            metrics_inc_replace_partition_number(partitions.len() as u64);
+            let vs = partitions
+                .into_iter()
+                .map(|partition| {
+                    let columns_min_max = partition
+                        .columns_min_max
+                        .into_iter()
+                        .map(|min_max| match min_max {
+                            MinMax::Point(v) => (v.clone(), v),
+                            MinMax::Range((min, max)) => (min, max),
+                        })
+                        .collect::<Vec<_>>();
+                    let key_hashes = partition.digests;
+                    DeletionByColumn {
+                        columns_min_max,
+                        key_hashes,
+                    }
+                })
+                .collect();
+            MergeIntoOperation::Delete(vs)
+        } else {
+            // otherwise, we just build a single delete action
+            self.build_merge_into_operation(&data_block_may_have_conflicts)?
+        };
+        Ok(merge_into_operation)
     }
 
-    fn extract_on_conflict_columns(
-        &mut self,
-        data_block: &DataBlock,
-    ) -> Result<MergeIntoOperation> {
+    fn table_level_row_prune(&self, data_block: &DataBlock) -> Result<DataBlock> {
+        let column_stats: &HashMap<ColumnId, ColumnStatistics> = &self.table_range_index;
+        let mut bitmap = MutableBitmap::new();
+        // for each row, check if it may have conflict
+        for row_idx in 0..data_block.num_rows() {
+            let mut should_keep = true;
+            // for each column, check if it may have conflict
+            for field in &self.on_conflict_fields {
+                let column: &Value<AnyType> = &data_block.columns()[field.field_index].value;
+                let value = column.row_scalar(row_idx)?;
+                let stats = column_stats.get(&field.table_field.column_id);
+                if let Some(stats) = stats {
+                    should_keep = !(value < stats.min.as_ref() || value > stats.max.as_ref());
+                    if !should_keep {
+                        // if one column not overlap,  no need to check other columns
+                        break;
+                    }
+                }
+            }
+            bitmap.push(should_keep);
+        }
+        let bitmap = bitmap.into();
+        data_block.clone().filter_with_bitmap(&bitmap)
+    }
```

Comment:
Can this loop be auto-vectorized? (By adjusting the inner and outer order of these two loops

Reply:
not sure :thinking: 

besides, it seems easier to maintain the  row bitmap, by letting the inner loop  break early if any one of the on-conflict key columns is out of table level range bound (indicates that this row will not trigger the `update` action)